### DISCUSSION TITLE
many to many with populateWhere

### DISCUSSION THREAD

**Discussion by yackinn** [mikro-orm/mikro-orm#2988] [Q&A]
Maybe we didn't configure the options correctly but we find some behavior that's unexpected.
The goal is to only populate where some condition is met without affecting the main resource.

```ts
// medium resource
this.repo.find(
        {
          createdBy: user.id
        },
        {
          populate: ['bookings'],
          fields: ['filename', 'alt'],
          populateWhere: { bookings: { createdBy: user.id } }
        }
```
When using `LoadStrategy.JOINED` it'll result in this sql leading to the wrong result.

```sql
SELECT
	"m0"."id",
	"m0"."filename",
	"m0"."alt",
	"b1"."id" AS "b1__id",
	"b1"."created_at" AS "b1__created_at", 	...
FROM
	"medium" AS "m0"
	LEFT JOIN "booking_media" AS "b2" ON "m0"."id" = "b2"."medium_id"
	LEFT JOIN "booking" AS "b1" ON "b2"."booking_id" = "b1"."id"
WHERE
	"m0"."created_by_id" = '5239ddb8-8daa-46ea-abee-e4b9ea8d5589'
```
This will include all bookings that are related to the medium inside the pivot table

When using `LoadStrategy.SELECT_IN` the result is correct. It yields all media entities that correspond to the where condition of the query `createdBy: user.id`. And it includes bookings on the media entities wherever the populateWhere condition matches to the booking entity.

That's the sql.

```sql
SELECT
	"m0"."id",
	"m0"."filename",
	"m0"."alt"
FROM
	"medium" AS "m0"
WHERE
	"m0"."created_by_id" = '5239ddb8-8daa-46ea-abee-e4b9ea8d5589';
	
SELECT
		"b0".*, "b1"."booking_id" AS "fk__booking_id", "b1"."medium_id" AS "fk__medium_id"
	FROM
		"booking" AS "b0"
	LEFT JOIN "booking_media" AS "b1" ON "b0"."id" = "b1"."booking_id"
WHERE
	"b0"."created_by_id" = '5239ddb8-8daa-46ea-abee-e4b9ea8d5589'
	AND "b1"."medium_id" in('60756680-e9b6-475c-a15c-39bbf2d0a07f', '0b67506d-0c64-4ef3-83cf-999f720dc177', ...)
```

Is this behavior correct? Do we have to use `LoadStrategy.SELECT_IN` for this to work?

Also using `populateWhere: PopulateHint.INFER` instead of `populateWhere: { bookings: { createdBy: user.id }}`
returns the wrong output with all bookings from the pivot table where it matches the medium id.

**Comment by B4nan** [comment]
Currently you need select-in strategy for this. Only way we could support this with joined strategy is wrapping in subqueries as with pagination support, I guess we might end up doing so in some future version, but right now `populateWhere` has no effect with joined strategy.

**Reply by yackinn** [reply]
Sure. Thank you.

**Reply by yackinn** [reply]
What's `PopulateHint.INFER` used for and when should you use it?

**Reply by B4nan** [reply]
It infers the where condition from your `em.find` call. In your case your condition is different so it produces different results.

```ts
this.repo.find(
        { createdBy: user.id }, // <-- where condition
        {
          populate: ['bookings'],
          fields: ['filename', 'alt'],
          populateWhere: PopulateHint.INFER // <-- same as `populateWhere: { createdBy: user.id }`
        }
```

**Reply by yackinn** [reply]
```ts
this.repo.find(
        { createdBy: user.id, bookings: { createdBy: user.id } }, // <-- where condition
        {
          populate: ['bookings'],
          fields: ['filename', 'alt'],
          populateWhere: PopulateHint.INFER // this should be `{bookings: { createdBy: user.id }}`
        }
```
This setup doesn't make sense correct?

We don't want the media results be restricted by the bookings condition so putting it in the where condition is ruled out I guess.

**Reply by B4nan** [reply]
That really depends on you what you want.

**Reply by yackinn** [reply]
The main resource should only be filtered by `createdBy: user.id`.
The booking entity should only be populated when `bookings: { createdBy: user.id }` matches.

Is there any way to use `populateWhere: PopulateHint.INFER`?

**Reply by B4nan** [reply]
I already told you how it works, not sure what more you'd like to hear. You don't want inferred condtion, so you can't use this hint.
